### PR TITLE
Composed component 

### PR DESCRIPTION
**Pull request by renancaraujo**
That is what a thought about Composed component''

For example of usage at https://github.com/renancaraujo/trex-flame: 
```dart
class HorizonLine extends PositionComponent with  Resizable, ComposedComponent {
  HorizonGround firstGround;
  HorizonGround secondGround;

  final double bumpThreshold = 0.5;

  HorizonLine(Image spriteImage){
    Sprite softSprite = Sprite.fromImage(spriteImage,
      width: HorizonDimensions.width,
      height: HorizonDimensions.height,
      y: 104.0,
      x: 2.0,
    );

    Sprite bumpySprite = Sprite.fromImage(spriteImage,
      width: HorizonDimensions.width,
      height: HorizonDimensions.height,
      y: 104.0,
      x: 2.0 + HorizonDimensions.width,
    );

    this.firstGround = HorizonGround(spriteImage, softSprite);
    this.secondGround = HorizonGround(spriteImage, bumpySprite);
    this..add(firstGround)..add(secondGround);
  }



  void update (t) {
    this.updateComponents((c){
      PositionComponent positionComponent = c as PositionComponent;
      positionComponent.y = y;
    });
   super.update(t);
  }
}


```

What you guys @luanpotter and @erickzanardo  think about that?

I'm willing to implement some tests after feedback.

### CODE REVIEW

**Comment by renancaraujo** [comment]
A thing that i want to discuss is the `iterateOverComponents`  method. 
I have a pretty bad taste for naming stuff, but basically I saw myself having to iterate over and over in the `components` set. Would you please give me some feedback about that? Thanks.

**Comment by luanpotter** [comment]
@renancaraujo , a couple points:

* if you implement (or extend, mixin, something something) `Resizable`, you can use the `children` method to return your components, and they will be resized as well (you don't actually need to do it yourself in the resize method). Probably you will need to keep the first time resize, though (might need take a look, as I don't quite recall right now).

* regarding the iterateOverComponents method, I'm fine with the naming; it's the utility that eludes me. You could just as easily do c.components.forEach(...), right? Is there any advantage to have that method? Note that you yourself have not used your implementation on the three instances you iterate over components (on render, update and resize).

Apart from that, I think this is an awesome addition and the implementation is pretty solid! Would love to merge :)

**Comment by renancaraujo** [comment]
Just pushed a commit addressing your considerations. About the `iterateOverComponents`: It was very useful only in my particular case, just stripped it out as soon as I realized that.

File: lib/components/composed_component.dart [hunk 1]
```diff
@@ -3,23 +3,23 @@ import 'dart:ui';
...
 import 'package:flame/components/resizable.dart';
+import 'package:flame/game.dart';
 import 'package:ordered_set/comparing.dart';
 import 'package:ordered_set/ordered_set.dart';
 
-
-typedef IterateOverComponents = void Function(Component component);
-
-abstract class ComposedComponent extends Component {
+/// A component that turns lets your component be composed others
```
Comment (luanpotter) [inline lib/components/composed_component.dart]: I think there is a small typo here ;)

File: lib/components/composed_component.dart [hunk 2]
```diff
@@ -43,17 +43,4 @@ abstract class ComposedComponent extends Component {
     }
   }
 
-  void iterateOverComponents ( IterateOverComponents iterateOverComponentsCallback ){
-    components.forEach(iterateOverComponentsCallback);
-  }
-
-  @override
```
Comment (luanpotter) [inline lib/components/composed_component.dart]: I think you should implement Resizable and implement the children method to return this.components, in order to keep this behavior. Like so:
```dart
  List<Resizable> children() => this.components.where((r) => r is Resizable).cast<Resizable>();
```
or something like that (that's from the top of my head, I have not tested it!)

**Comment by renancaraujo** [comment]
@luanpotter Made some updates, including items pointed by your review and the supermixin syntax (`mixin` `on`)

**Comment by renancaraujo** [comment]
The actual changes works perfectly with Resizable. ref: 

https://github.com/renancaraujo/trex-flame/blob/b80c4d9beea4e2009d3f2d9e9f2fe429f4274107/lib/game/custom/composed_component.dart

https://github.com/renancaraujo/trex-flame/blob/b80c4d9beea4e2009d3f2d9e9f2fe429f4274107/lib/game/t_rex/t_rex.dart

**Comment by luanpotter** [comment]
Thanks <3

**Pull request merged**